src/bun.js/bindings/NodeTLS.cpp

```diff
@@ -6,35 +6,66 @@
 #include "libusockets.h"
 
 #include "ZigGlobalObject.h"
+#include "openssl/base.h"
+#include "openssl/bio.h"
+#include "../../packages/bun-usockets/src/crypto/root_certs_header.h"
 
 namespace Bun {
 
 using namespace JSC;
 
-BUN_DECLARE_HOST_FUNCTION(Bun__canonicalizeIP);
-
-JSC::JSValue createNodeTLSBinding(Zig::GlobalObject* globalObject)
+JSC_DEFINE_HOST_FUNCTION(getBundledRootCertificates, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
 {
     VM& vm = globalObject->vm();
-    auto* obj = constructEmptyObject(globalObject);
 
     struct us_cert_string_t* out;
     auto size = us_raw_root_certs(&out);
     if (size < 0) {
-        return jsUndefined();
+        return JSValue::encode(jsUndefined());
     }
     auto rootCertificates = JSC::JSArray::create(vm, globalObject->arrayStructureForIndexingTypeDuringAllocation(JSC::ArrayWithContiguous), size);
     for (auto i = 0; i < size; i++) {
         auto raw = out[i];
         auto str = WTF::String::fromUTF8(std::span { raw.str, raw.len });
         rootCertificates->putDirectIndex(globalObject, i, JSC::jsString(vm, str));
     }
-    obj->putDirect(
-        vm, JSC::PropertyName(JSC::Identifier::fromString(vm, "rootCertificates"_s)), JSC::objectConstructorFreeze(globalObject, rootCertificates), 0);
 
-    obj->putDirect(
-        vm, JSC::PropertyName(JSC::Identifier::fromString(vm, "canonicalizeIP"_s)), JSC::JSFunction::create(vm, globalObject, 1, "canonicalizeIP"_s, Bun__canonicalizeIP, ImplementationVisibility::Public, NoIntrinsic), 0);
-    return obj;
+    return JSValue::encode(JSC::objectConstructorFreeze(globalObject, rootCertificates));
+}
+
+JSC_DEFINE_HOST_FUNCTION(getExtraCACertificates, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    auto scope = DECLARE_THROW_SCOPE(globalObject->vm());
+    VM& vm = globalObject->vm();
+
+    STACK_OF(X509)* root_extra_cert_instances = us_get_root_extra_cert_instances();
+
+    auto size = sk_X509_num(root_extra_cert_instances);
+    if (size < 0) {
+        return JSValue::encode(jsUndefined());
+    }
+
+    auto rootCertificates = JSC::JSArray::create(vm, globalObject->arrayStructureForIndexingTypeDuringAllocation(JSC::ArrayWithContiguous), size);
+    for (auto i = 0; i < size; i++) {
+        BIO* bio = BIO_new(BIO_s_mem());
+        if (!bio) {
+            throwOutOfMemoryError(globalObject, scope);
+            return {};
+        }
+
+        if (PEM_write_bio_X509(bio, sk_X509_value(root_extra_cert_instances, i)) != 1) {
+            BIO_free(bio);
+            continue;
+        }
+
+        char* bioData = nullptr;
+        size_t bioLength = BIO_get_mem_data(bio, &bioData);
```

Comment:
I haven't checked BoringSSL, but I think we should have a similar length check to Node
```
auto pem_size = BIO_get_mem_data(pem_bio.get(), &pem_data);
if (pem_size <= 0 || !pem_data) {
      ThrowCryptoError(env, ERR_get_error(), "Reading PEM data");
      return MaybeLocal<Array>();
    }
```